t/t5799-gvfs-helper.sh

```diff
@@ -0,0 +1,361 @@
+# them and so that we don't have to hard-code a list of known OIDs.
+# This doesn't need to be a complete list -- just enough to drive some
+# representative tests.
+#
+# Optionally require that we find a minimum number of OIDs.
+#
+get_list_of_oids() {
+	git -C "$REPO_SRC" rev-list --objects HEAD | sed 's/ .*//' | sort >"$OIDS_FILE"
+
+	if test $# -eq 1
+	then
+		if ! test $(wc -l <"$OIDS_FILE") -ge $1
+		then
+			echo "get_list_of_oids: insufficient data.  Need $1 OIDs."
+			return 1
+		fi
+	fi
+	return 0
+}
+
+test_expect_success 'setup repos' '
+	test_create_repo "$REPO_SRC" &&
+	#
+	# test_commit_bulk() does magic to create a packfile containing
+	# the new commits.
+	#
+	test_commit_bulk -C "$REPO_SRC" --filename="batch_a.%s.t" 9 &&
+	test_commit_bulk -C "$REPO_SRC" --filename="batch_b.%s.t" 9 &&
+	#
+	# test_commit() creates commits, trees, tags, and blobs and leave
+	# them loose.
+	#
+	test_commit -C "$REPO_SRC" file1.txt &&
+	test_commit -C "$REPO_SRC" file2.txt &&
+	test_commit -C "$REPO_SRC" file3.txt &&
+	test_commit -C "$REPO_SRC" file4.txt &&
+	test_commit -C "$REPO_SRC" file5.txt &&
+	test_commit -C "$REPO_SRC" file6.txt &&
+	test_commit -C "$REPO_SRC" file7.txt &&
+	test_commit -C "$REPO_SRC" file8.txt &&
+	test_commit -C "$REPO_SRC" file9.txt &&
+	#
+	# gvfs-helper.exe writes downloaded objects to a shared-cache directory
+	# rather than the ODB inside the .git directory.
+	#
+	mkdir "$SHARED_CACHE_T1" &&
+	#
+	# setup repo_t1 and point all of the gvfs.* values to repo_src.
+	#
+	test_create_repo "$REPO_T1" &&
+	git -C "$REPO_T1" remote add origin $ORIGIN_URL &&
+	git -C "$REPO_T1" config --local gvfs.cache-server $CACHE_URL &&
+	git -C "$REPO_T1" config --local gvfs.sharedcache "$SHARED_CACHE_T1" &&
+	echo "$SHARED_CACHE_T1" >> "$REPO_T1"/.git/objects/info/alternates &&
+	#
+	# Create some test data sets.
+	#
+	get_list_of_oids 30
+'
+
+stop_gvfs_protocol_server() {
+	if ! test -f "$PID_FILE"
+	then
+		return 0
+	fi
+	#
+	# The server will shutdown automatically when we delete the pid-file.
+	#
+	rm -f "$PID_FILE"
+	#
+	# Give it a few seconds to shutdown (mainly to completely release the
+	# port before the next test start another instance and it attempts to
+	# bind to it).
+	#
+	for k in 0 1 2 3 4
+	do
+		if grep -q "Starting graceful shutdown" "$SERVER_LOG"
+		then
+			return 0
+		fi
+		sleep 1
+	done
+
+	echo "stop_gvfs_protocol_server: timeout waiting for server shutdown"
+	return 1
+}
+
+start_gvfs_protocol_server() {
+	#
+	# Launch our server into the background in repo_src.
+	#
+	(
+		cd "$REPO_SRC"
+		test-gvfs-protocol --verbose \
+			--listen=127.0.0.1 \
+			--port=$GIT_TEST_GVFS_PROTOCOL_PORT \
+			--reuseaddr \
+			--pid-file="$PID_FILE" \
+			2>"$SERVER_LOG" &
+	)
+	#
+	# Give it a few seconds to get started.
+	#
+	for k in 0 1 2 3 4
+	do
+		if [ -f "$PID_FILE" ]
+		then
+			return 0
+		fi
+		sleep 1
+	done
+
+	echo "start_gvfs_protocol_server: timeout waiting for server startup"
```

Comment:
I want to start and stop the server for each test -- both for test isolation and so that i can change the startup args.  So I put the stop_ call in the test_when_finished helper.